File: backend/NalaCreditAPI/Models/InterBranchTransfer.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace NalaCreditAPI.Models;

public class InterBranchTransfer
{
    public Guid Id { get; set; } = Guid.NewGuid();

    [Required]
    [MaxLength(50)]
    public string TransferNumber { get; set; } = string.Empty;

    // Source Branch
    [Required]
    public int FromBranchId { get; set; }
    [ForeignKey("FromBranchId")]
    public Branch FromBranch { get; set; } = null!;

    [Required]
    [MaxLength(100)]
    public string FromBranchName { get; set; } = string.Empty;

    // Destination Branch
    [Required]
    public int ToBranchId { get; set; }
    [ForeignKey("ToBranchId")]
    public Branch ToBranch { get; set; } = null!;

    [Required]
    [MaxLength(100)]
    public string ToBranchName { get; set; } = string.Empty;

    // Transfer Details
    [Required]
    public ClientCurrency Currency { get; set; }

    [Required]
    [Range(0.01, double.MaxValue)]
    public decimal Amount { get; set; }

    [Required]
    [Range(0.01, double.MaxValue)]
    public decimal ExchangeRate { get; set; } = 1.0m;

    public decimal ConvertedAmount { get; set; }

    [Required]
    [MaxLength(500)]
    public string Reason { get; set; } = string.Empty;

    [MaxLength(1000)]
    public string? Notes { get; set; }

    // Status and Approval
    [Required]
    public TransferStatus Status { get; set; } = TransferStatus.Pending;

    [Required]
    [MaxLength(100)]
    public string RequestedBy { get; set; } = string.Empty;

    public string? RequestedByName { get; set; }

    public DateTime? RequestedAt { get; set; }

    public string? ApprovedBy { get; set; }
    public string? ApprovedByName { get; set; }
    public DateTime? ApprovedAt { get; set; }

    public string? RejectedBy { get; set; }
    public string? RejectedByName { get; set; }
    public string? RejectionReason { get; set; }
    public DateTime? RejectedAt { get; set; }

    public string? ProcessedBy { get; set; }
    public string? ProcessedByName { get; set; }
    public DateTime? ProcessedAt { get; set; }

    // Tracking
    [MaxLength(100)]
    public string? ReferenceNumber { get; set; }

    [MaxLength(100)]
    public string? TrackingNumber { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    // Navigation properties
    public ICollection<InterBranchTransferLog> TransferLogs { get; set; } = new List<InterBranchTransferLog>();
}

public enum TransferStatus
{
    Pending = 0,
    Approved = 1,
    Rejected = 2,
    InTransit = 3,
    Completed = 4,
    Cancelled = 5
}

public class InterBranchTransferLog
{
    public Guid Id { get; set; } = Guid.NewGuid();

    [Required]
    public Guid TransferId { get; set; }
    [ForeignKey("TransferId")]
    public InterBranchTransfer Transfer { get; set; } = null!;

    [Required]
    [MaxLength(100)]
    public string Action { get; set; } = string.Empty;

    [Required]
    [MaxLength(500)]
    public string Description { get; set; } = string.Empty;

    [Required]
    [MaxLength(100)]
    public string PerformedBy { get; set; } = string.Empty;

    public string? PerformedByName { get; set; }

    public DateTime PerformedAt { get; set; } = DateTime.UtcNow;

    public string? OldValue { get; set; }
    public string? NewValue { get; set; }
}

File: frontend-web/src/types/interBranchTransfer.ts
export interface InterBranchTransfer {
  id: string;
  transferNumber: string;
  fromBranchId: number;
  fromBranchName: string;
  toBranchId: number;
  toBranchName: string;
  currency: Currency;
  currencyName: string;
  amount: number;
  exchangeRate: number;
  convertedAmount: number;
  reason: string;
  notes?: string;
  status: TransferStatus;
  statusName: string;
  requestedBy: string;
  requestedByName?: string;
  requestedAt?: string;
  approvedBy?: string;
  approvedByName?: string;
  approvedAt?: string;
  rejectedBy?: string;
  rejectedByName?: string;
  rejectionReason?: string;
  rejectedAt?: string;
  processedBy?: string;
  processedByName?: string;
  processedAt?: string;
  referenceNumber?: string;
  trackingNumber?: string;
  createdAt: string;
  updatedAt: string;
}

export enum Currency {
  HTG = 0,
  USD = 1
}

export enum TransferStatus {
  Pending = 0,
  Approved = 1,
  Rejected = 2,
  InTransit = 3,
  Completed = 4,
  Cancelled = 5
}

export interface CreateInterBranchTransferDto {
  fromBranchId?: number; // Optional for SuperAdmin, required if user has no BranchId
  toBranchId: number;
  currency: Currency;
  amount: number;
  exchangeRate?: number;
  reason: string;
  notes?: string;
}

export interface UpdateInterBranchTransferDto {
  id: string;
  amount?: number;
  exchangeRate?: number;
  reason?: string;
  notes?: string;
}

export interface ApproveInterBranchTransferDto {
  id: string;
  notes?: string;
}

export interface RejectInterBranchTransferDto {
  id: string;
  reason: string;
}

export interface ProcessInterBranchTransferDto {
  id: string;
  referenceNumber?: string;
  trackingNumber?: string;
  notes?: string;
}

export interface DispatchInterBranchTransferDto {
  id: string;
  referenceNumber?: string;
  trackingNumber?: string;
  notes?: string;
}

export interface InterBranchTransferSearchDto {
  fromBranchId?: number;
  toBranchId?: number;
  currency?: Currency;
  status?: TransferStatus;
  requestedBy?: string;
  startDate?: string;
  endDate?: string;
  page?: number;
  pageSize?: number;
}

export interface InterBranchTransferLogDto {
  id: string;
  transferId: string;
  action: string;
  description: string;
  performedBy: string;
  performedByName?: string;
  performedAt: string;
  oldValue?: any;
  newValue?: any;
}

export interface BranchTransferSummaryDto {
  branchId: number;
  branchName: string;
  totalSent: number;
  totalReceived: number;
  totalSentHTG?: number;
  totalReceivedHTG?: number;
  totalSentUSD?: number;
  totalReceivedUSD?: number;
  pendingTransfers: number;
  completedTransfers: number;
  lastTransferDate: string;
}

export interface ConsolidatedTransferReportDto {
  branchSummaries: BranchTransferSummaryDto[];
  totalSystemTransfers: number;
  totalActiveTransfers: number;
  reportGeneratedAt: string;
}

// Form data interfaces for frontend components
export interface InterBranchTransferFormData {
  fromBranchId?: number; // Optional for users with BranchId, required for SuperAdmin
  toBranchId: number;
  currency: Currency;
  amount: number;
  exchangeRate?: number;
  reason: string;
  notes?: string;
}

export interface TransferApprovalFormData {
  notes?: string;
}

export interface TransferRejectionFormData {
  reason: string;
}

export interface TransferProcessingFormData {
  referenceNumber?: string;
  trackingNumber?: string;
  notes?: string;
}

// Status display helpers
export const getTransferStatusInfo = (status: TransferStatus) => {
  const statusMap = {
    [TransferStatus.Pending]: {
      label: 'En attente',
      color: 'bg-yellow-100 text-yellow-800',
      icon: 'Clock'
    },
    [TransferStatus.Approved]: {
      label: 'Approuvé',
      color: 'bg-blue-100 text-blue-800',
      icon: 'CheckCircle'
    },
    [TransferStatus.InTransit]: {
      label: 'En transit',
      color: 'bg-purple-100 text-purple-800',
      icon: 'Truck'
    },
    [TransferStatus.Completed]: {
      label: 'Terminé',
      color: 'bg-green-100 text-green-800',
      icon: 'CheckCircle2'
    },
    [TransferStatus.Rejected]: {
      label: 'Rejeté',
      color: 'bg-red-100 text-red-800',
      icon: 'XCircle'
    },
    [TransferStatus.Cancelled]: {
      label: 'Annulé',
      color: 'bg-gray-100 text-gray-800',
      icon: 'X'
    }
  };
  return statusMap[status] || statusMap[TransferStatus.Pending];
};

export const getCurrencyInfo = (currency: Currency) => {
  const currencyMap = {
    [Currency.HTG]: {
      code: 'HTG',
      name: 'Gourdes Haitiennes',
      symbol: 'HTG'
    },
    [Currency.USD]: {
      code: 'USD',
      name: 'US Dollar',
      symbol: '$'
    }
  };
  return currencyMap[currency] || currencyMap[Currency.HTG];
};

// Action types for transfer workflow
export enum TransferAction {
  Create = 'Create',
  Approve = 'Approve',
  Reject = 'Reject',
  Process = 'Process',
  Cancel = 'Cancel'
}

// Transfer workflow configuration
export const TRANSFER_WORKFLOW = {
  [TransferStatus.Pending]: {
    allowedActions: [TransferAction.Approve, TransferAction.Reject],
    nextStatuses: [TransferStatus.Approved, TransferStatus.Rejected]
  },
  [TransferStatus.Approved]: {
    allowedActions: [TransferAction.Process, TransferAction.Cancel],
    nextStatuses: [TransferStatus.InTransit, TransferStatus.Cancelled]
  },
  [TransferStatus.InTransit]: {
    allowedActions: [TransferAction.Process],
    nextStatuses: [TransferStatus.Completed]
  },
  [TransferStatus.Completed]: {
    allowedActions: [],
    nextStatuses: []
  },
  [TransferStatus.Rejected]: {
    allowedActions: [],
    nextStatuses: []
  },
  [TransferStatus.Cancelled]: {
    allowedActions: [],
    nextStatuses: []
  }
};